data/transactions/logic/assembler.go

```diff
@@ -1666,6 +1681,80 @@ func (ops *OpStream) assemble(text string) error {
 	return nil
 }
 
+func (ops *OpStream) cycle(macro string, previous ...string) bool {
+	replacement, ok := ops.macros[macro]
+	if !ok {
+		return false
+	}
+	if len(previous) > 0 && macro == previous[0] {
+		ops.errorf("Macro cycle discovered: %s", strings.Join(append(previous, macro), " -> "))
+		return true
+	}
+	for _, token := range replacement {
+		if ops.cycle(token, append(previous, macro)...) {
+			return true
+		}
+	}
+	return false
+}
+
+// Once a version is obtained, versionedMacroFilter goes through previously defined macros
+// and ensures they don't use opcodes/fields from the obtained version
+func (ops *OpStream) versionedMacroFilter() error {
+	var err error = nil
+	for macroName := range ops.macros {
+		if fieldNames[ops.Version][macroName] {
+			err = fmt.Errorf("%s is defined as a macro but is a field name in version %d", macroName, ops.Version)
+		} else if _, ok := OpsByName[ops.Version][macroName]; ok {
+			err = fmt.Errorf("%s is defined as a macro but is an opcode in version %d", macroName, ops.Version)
+		} else {
+			continue
+		}
+		ops.error(err)
+		delete(ops.macros, macroName)
+	}
+	if err != nil {
+		err = errors.New("version is incompatible with defined macros")
+	}
+	return err
+}
+
+func (ops *OpStream) filterMacro(macroName string) error {
+	_, isTxnType := txnTypeMap[macroName]
+	_, isOnCompletion := onCompletionMap[macroName]
+	if isTxnType || isOnCompletion {
+		return ops.errorf("Named constants cannot be used as macro names: %s", macroName)
+	}
+	if ops.Version != assemblerNoVersion {
+		if _, ok := OpsByName[ops.Version][macroName]; ok {
+			return ops.errorf("Macro names cannot be opcodes: %s", macroName)
+		}
+		if fieldNames[ops.Version][macroName] {
+			return ops.errorf("Macro names cannot be field names: %s", macroName)
+		}
+	}
+	return nil
+}
+
+func (ops *OpStream) define(tokens []string) error {
+	if tokens[0] != "#define" {
+		return ops.errorf("invalid syntax: %s", tokens[0])
+	}
+	if len(tokens) < 3 {
+		return ops.errorf("define directive requires a name and body")
+	}
+	saved, ok := ops.macros[tokens[1]]
```

Comment:
What if I already defined a macro with the same name? I don't see that case handled here

Reply:
We want you to be able to redefine macros so if you do:
#define rowsize 20
... some stuff
#define rowsize 30
... more stuff
rowsize will be a valid macro that is different depending on where it's used

Reply:
Ok great, I'll resolve this